### PR TITLE
Initialize destinations for destination REST API

### PR DESCRIPTION
**Pull request by jsantos17**
The most relevant changes are in `quasar.run.Quasar`. There are other changes that are necessary for subsequent PRs with the actual API implementation.

Part of [ch7014]

Changes are `breaking` because of changes in `quasar.run.Quasar.apply`.

### CODE REVIEW

**Comment by jsantos17** [comment]
Released as `128.1.0-107524a`

**Comment by jsantos17** [comment]
🛑up since I'll need to make changes in this branch for dependent PR reviews.

File: impl/src/main/scala/quasar/impl/destinations/DefaultDestinations.scala [hunk 1]
```diff
@@ -34,7 +34,7 @@ import scalaz.syntax.monad._
 import scalaz.{\/, Equal, IMap, ISet, OptionT, Order}
 import shims._
 
-class DefaultDestinations[F[_]: Sync, I: Equal: Order, C] private (
+class DefaultDestinations[I: Equal: Order, C, F[_]: Sync] private (
```
Comment (cryogenian) [inline impl/src/main/scala/quasar/impl/destinations/DefaultDestinations.scala]: Why do you prefer this?

File: api/src/main/scala/quasar/api/datasource/DatasourceType.scala [hunk 2]
```diff
@@ -16,23 +16,34 @@
...
 import scalaz.syntax.show._
+import shapeless.{Witness => W}
 
 @Lenses
 final case class DatasourceType(name: DatasourceType.Name, version: Positive)
 
 object DatasourceType extends DatasourceTypeInstances {
-  type Name = datasource.Name
-  type NameP = datasource.NameP
+  type NameP = MatchesRegex[W.`"[a-zA-Z0-9-]+"`.T]
```
Comment (cryogenian) [inline api/src/main/scala/quasar/api/datasource/DatasourceType.scala]: Why should this be defined here? Just wondering :)

File: impl/src/main/scala/quasar/impl/destinations/DefaultDestinations.scala (hunk 1 above)
Comment (jsantos17) [reply impl/src/main/scala/quasar/impl/destinations/DefaultDestinations.scala]: I wanted to keep it consistent with `DefaultDestinationManager`. `F[_]` at the end won.

File: api/src/main/scala/quasar/api/datasource/DatasourceType.scala (hunk 2 above)
Comment (jsantos17) [reply api/src/main/scala/quasar/api/datasource/DatasourceType.scala]: No good reason. For some reason they aren't found when put in `quasar.api`'s package object :/

**Comment by jsantos17** [comment]
Closing in favor of #4202

**Pull request closed without merging**